File: tp_tda/tp2/bipartito.py
from collections import deque

def adjacencyList(edgeList):
    adjacencyList = {}
    for node1, node2 in edgeList:
        if node1 not in adjacencyList:
            adjacencyList[node1] = []
        if node2 not in adjacencyList:
            adjacencyList[node2] = []
        adjacencyList[node1].append(node2)
        adjacencyList[node2].append(node1)
    return adjacencyList

def bfs(adjList, start):
    queue = deque([start]) 
    visited = {start: True}
    level = {start: 0}
              
    evenVertices = []
    oddVertices = []
    
    while queue:
        currentNode = queue.popleft()
        
        if level[currentNode] % 2 == 0:
            evenVertices.append(currentNode)
        else:
            oddVertices.append(currentNode)
        
        for neighbor in adjList[currentNode]:
            if neighbor not in visited:   
                queue.append(neighbor)
                level[neighbor] = level[currentNode] + 1
                visited[neighbor] = True
                    
    return evenVertices, oddVertices

def main():
    with open("input.txt", "r") as archive:
        lines = archive.readlines()

    vertices = int(lines[0].strip())

    edges = []
    for line in lines[1:]:
        u, v = map(int, line.split())
        edges.append((u, v))

    adjList = adjacencyList(edges)
    
    if len(edges) == 0 or vertices <= 1:
        print(0)
    else:
        start_node = list(adjList.keys())[0]
        a,b = bfs(adjList, start_node)
    
        print((len(a)*len(b) - len(edges)))
    
main()

File: tp_tda/tp2/bipartito.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>

using namespace std;

unordered_map<int, vector<int>> adjacencyList(const vector<pair<int, int>>& edgeList) {
    unordered_map<int, vector<int>> adjList;
    for (const auto& edge : edgeList) {
        int node1 = edge.first, node2 = edge.second;
        adjList[node1].push_back(node2);
        adjList[node2].push_back(node1);
    }
    return adjList;
}

pair<int, int> bfs(const unordered_map<int, vector<int>>& adjList, int start) {
    queue<int> q;
    unordered_map<int, bool> visited;
    unordered_map<int, int> level;
    
    q.push(start);
    visited[start] = true;
    level[start] = 0;
    
    int evenVertices = 0, oddVertices = 0;
    
    while (!q.empty()) {
        int currentNode = q.front();
        q.pop();
        
        if (level[currentNode] % 2 == 0) {
            evenVertices++;
        } else {
            oddVertices++;
        }
        
        for (int neighbor : adjList.at(currentNode)) {
            if (!visited[neighbor]) {
                q.push(neighbor);
                level[neighbor] = level[currentNode] + 1;
                visited[neighbor] = true;
            }
        }
    }
    
    return {evenVertices, oddVertices};
}

int main() {
    int vertices;
    cin >> vertices;
    
    vector<pair<int, int>> edges;
    for (int i = 0; i < vertices - 1; ++i) {
        int u, v;
        cin >> u >> v;
        edges.emplace_back(u, v);
    }
    
    auto adjList = adjacencyList(edges);
    
    if (edges.empty() || vertices <= 1) {
        cout << 0 << endl;
    } else {
        int start_node = edges[0].first;
        auto [evenCount, oddCount] = bfs(adjList, start_node);
        cout << (evenCount * oddCount - edges.size()) << endl;
    }
    
    return 0;
}
